### PR TITLE
Create a CupertinoApp

### PR DESCRIPTION
**Pull request by xster**
Part 1 of #18037 

- Back port some doc updates from MaterialApp to WidgetsApp
- Extract common parts to doc templates to prevent future drift
- Add a CupertinoApp (that reuses CupertinoTabView instead of re-implementing navigator routing yet one more time)
- Clean all the tests

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/app.dart [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+  void _updateNavigator() {
+    _haveNavigator = widget.home != null ||
+                     widget.routes.isNotEmpty ||
+                     widget.onGenerateRoute != null ||
+                     widget.onUnknownRoute != null;
+  }
+
+  // The `child` coming back out from WidgetsApp will always be null since
+  // we never passed in anything for it to create a Navigator inside
+  // WidgetsApp.
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/app.dart]: add an assert to check that

File: packages/flutter/lib/src/material/app.dart [hunk 2]
```diff
@@ -218,164 +218,48 @@ class MaterialApp extends StatefulWidget {
   /// [builder] must not be null.
   final Map<String, WidgetBuilder> routes;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/app.dart]: iirc, some of the docs in this file were subtly different than those in widgets/app.dart in various minor ways. Please verify that the ones you've removed are in fact the same byte-for-byte.

**Comment by Hixie** [comment]
can we get away with not including the Navigator in CupertinoApp at all?

File: packages/flutter/lib/src/cupertino/app.dart (hunk 1 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/app.dart]: Done. Added some comments here and below too about not letting WidgetsApp make a navigator.

**Comment by xster** [comment]
In the current state, for consistency with MaterialApp, no :(
Users don't necessarily want to use bottom tabs which also provides navigators. 

Though I think if you extract a router out of the MaterialApp in a future refactor, it's worth removing it from here as well.

File: packages/flutter/lib/src/material/app.dart (hunk 2 above)
Comment (xster) [reply packages/flutter/lib/src/material/app.dart]: Reverted some unintended merging of MaterialApp's doc into WidgetsApp. Mainly references to `routes`. Double checked that the delete docs from MaterialApp appear word for word in the template.

**Comment by Hixie** [comment]
LGTM

**Comment by xster** [comment]
Waiting for the engine roll (which has a golden update queued up) before merging.

**Pull request merged**